### DISCUSSION TITLE
How to define channels for ti,ads1x1x-base

### DISCUSSION THREAD

**Discussion by naNEQ** [zephyrproject-rtos/zephyr#61960] [Q&A]
I've been trying to enable the ADC driver to use it with my external ADS1015 IC, but I haven't been able to succeed so far. The sample application is unfortunately only using the internal ADCs of the MCUs, so I don't have a good reference on how to do it.

I'm targeting an ESP32 custom board.

First things first, this is how my overlay looked like in the beginning:

```C
&i2c0 {
	lis2dh@18 {
		compatible = "st,lis2dh";
		label = "LIS2DH";
		reg = <0x18>;
		irq-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>, <&gpio0 30 GPIO_ACTIVE_HIGH>;
		disconnect-sdo-sa0-pull-up;
	};
	ads1015@48 {
		compatible = "ti,ads1015";
		#io-channel-cells = <1>;
		reg = <0x48 >;
	};
};
```

So far so good, I could get the device using the DEVICE_DT_GET_ANY

```C
// Initialize an ADS1015 driver
const struct device *adc_dev = DEVICE_DT_GET_ANY(ti_ads1015);
if (adc_dev == NULL)
{
	LOG_ERR("Failed to get ADS1015 ADC device binding\n");
	return;
}
```

But then I have to be able to configure the individual channels of the ADC, so I tried doing it how it is shown in the sample for the ADC driver. So I thought I should add the channel subnode below the ADC

```C
&i2c0 {
	lis2dh@18 {
		compatible = "st,lis2dh";
		label = "LIS2DH";
		reg = <0x18>;
		irq-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>, <&gpio0 30 GPIO_ACTIVE_HIGH>;
		disconnect-sdo-sa0-pull-up;
	};
	ads1015@48 {
		compatible = "ti,ads1015";
		#io-channel-cells = <1>;
		reg = <0x48 >;

		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <12>;
		};
	};
};
```

But I have the feeling that this is not the correct way, because I then get a weird devicetree error

```C
devicetree error: 'reg' property in <Node /soc/i2c@3ff53000/ads1015@48/channel@0 in 'C:/Users/NikosAgianniotis/hvac-iot/zephyr/misc/empty_file.c'> has length 4, which is not evenly divisible by 12 (= 4*(<#address-cells> (= 2) + <#size-cells> (= 1))). Note that #*-cells properties come either from the parent node or from the controller (in the case of 'interrupts').
```

Has anybody tried it or has a good bet on how to do it properly?

**Comment by naNEQ** [comment]
To reply my own question, I think I found the answer. The following worked just perfect (at least I can build, I haven't tried it on a real hardware to see if it works yet) and if I understand it correctly, I think the problem was that I was missing the two parameters "#address-cells" and "#size-cells".

```C
&i2c0 {
	lis2dh@18 {
		compatible = "st,lis2dh";
		label = "LIS2DH";
		reg = <0x18>;
		irq-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>, <&gpio0 30 GPIO_ACTIVE_HIGH>;
		disconnect-sdo-sa0-pull-up;
	};
	i2c_adc: ads1015@48 {
		compatible = "ti,ads1015";
		reg = <0x48>;
		#io-channel-cells = <1>;
		#address-cells = <1>;
		#size-cells = <0>;
		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_VDD_1";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		};
		channel@1 {
			reg = <1>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_VDD_1";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		};
		channel@2 {
			reg = <2>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_VDD_1";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		};
		channel@3 {
			reg = <3>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_VDD_1";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		};
	};
};
```

**Comment by hubert-tt** [comment]
Hi, I had a similar problem with another ADC module and your solution helped me - the project is building. I'd like to know if you've managed to use this with real hardware and if so, how did you refer to the individual channels of the external ADC in the code, because I'm having trouble with this?

**Reply by naNEQ** [reply]
Hi,

I'm happy my post helped you.

The thing is that in the end I abandoned the idea of using this driver and just wrote my own library that accesses the ADC. In my opinion the driver for this specific ADC is a bit cumbersome to work with. The main thing I realized was that in reality the driver treats the ADC as an one-channel ADC. At least this is what I understood by going deep into the driver source code. I think additional channels, other than the first one, are practically ignored.

Therefore the only way to get samples from all channels is to create different adc_channel_cfg structures and change the input_positive parameter to reflect the number of MUX input you want.

But I'd be interested to know if you have a different understanding of how the driver is structured and if you manage to make it work.

**Reply by hubert-tt** [reply]
I was using different external ADC, MCP3204 in my case, and yesterday I've managed to read data from all of its channels doing everything step by step, because I couldn't use `zephyr,user` and `io-channels` property to simplify it.

**Reply by naNEQ** [reply]
Happy to hear that! Can you clarify a bit what you mean by "step by step"? Did you also use your own library?

**Reply by hubert-tt** [reply]
No, I used driver from Zephyr, but I have to manually fetch all channels data, setup the device etc., not just by using `adc_dt_spec` structure and functions associated with it.

**Comment by naNEQ** [comment]
Got it, thanks

**Comment by GilDev** [comment]
Hi, have the same issue with an ADS1115. I configured everything like you but the ADC shell command doesn’t list the ADC (only the internal one), not sure how to proceed from here, there must be a way…

**Reply by naNEQ** [reply]
It is really cumbersome to work with the driver. And as I wrote some months ago, the driver treats the ADC as one channel ADC, which is not exactly true.

I personally gave up and made my own version of it. It's not a proper Zephyr driver, but rather a library. But it seems to work.

**Reply by GilDev** [reply]
Okay I finally made it work, for reference here is my device tree overlay:

```
/ {
	zephyr,user {
		io-channels = <&ads1115 0>;
	};
};

&i2c2 {
	status = "okay";
	ads1115: ads1115@48 {
		status = "okay";
		compatible = "ti,ads1115";
		reg = <0x48>;
		#io-channel-cells = < 1 >;
		#address-cells = < 1 >;
		#size-cells = < 0 >;

		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
		};
		channel@1 {
			reg = <1>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
		};
		channel@2 {
			reg = <2>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
		};
		channel@3 {
			reg = <3>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
		};
	};
};
```

But indeed only channel 0 is supported, that is SO weird… I guess I’m also gonna have to DIY a library on the side for now…

For whatever reason it also doesn’t show up when typing the `adc` command in the shell…

**Reply by GastonMelo** [reply]
@GilDev Im using the ads1112 from TI and I want to enable the driver n zephyr, but there is no example or something else. Reading the Kconfig of this device I can see that it needs CONFIG_ADC=y and also depends on DT_HAS_TI_ADC_ADS1112_ENABLED....where do I set this devicetree entry? in the source file?

**Reply by GilDev** [reply]
I ended up also writing my own basic library. This driver is not complete and Zephyr has a too steep learning curve for me to do it better right now sadly…

**Comment by hubert-tt** [comment]
DT_HAS_TI_ADC_ADS1112_ENABLED is set when you add node compatible with it
to your device tree overlay and set its status to "okay"

czw., 19 wrz 2024, 20:58 użytkownik Gaston ***@***.***>
napisał:

**Reply by GastonMelo** [reply]
@hubert-tt  thansk for ur help --- this is really frustating and I'm just starting with zephyr , and i see a long long road ahead of me.
I had a i2c0 node in the overlay file, in there ii have 2 more devices (oled and a camera) for testing purpose only. An strange error is happening, if I add the node to the i2c (the adc uses the i2c for coms):

`
&i2c0 {
	ads1100: ads1100@48 {
	compatible = "ti,ads1100";
        status = "okay";
	reg = <0x48>;
    };    
    ov7670_camera: ov7670_camera@21 {
        compatible= "ovti,ov7670";
        status = "okay";
        reg=<0x21>;
        reset-gpios = <&gpio0 2 GPIO_ACTIVE_HIGH>;
    };  
        oled0: oled0@3c {
        compatible= "solomon,ssd1306fb";
        reg =<0x3c>;
        height=<64>;
        width=<128>;
        segment-offset=<0>;
        page-offset=<0>;
        display-offset=<0>;
        multiplex-ratio=<63>;
        segment-remap;
        com-invdir;
        prechargep= <0x22>;
    };     
};
`
i got and error saying:

`
/home/gaston/Zephyrprojects/external/zephyr/drivers/adc/ads1100/ads1100.c:12:10: fatal error: adc_context.h: No such file or directory
   12 | #include "adc_context.h"
      |          ^~~~~~~~~~~~~~~
`
If I remove the ads1100 node from the i2c0 node I can compile the code without errors.
Sorry to ask this,  but I just start with zyphr.
I don't understand what is causing this error, as far I understand i need to define a node inside the i2c0 device.
Regards
Gaston

**Reply by naNEQ** [reply]
@GastonMelo I suppose you've made your own external out-of-the-tree driver? You're just missing this file the error message is mentioning. This has to be present in the same directory as your `ads1100.c` file. Just take it from in-tree. You have to do it like this, because it is not included in a default include directory.

At least this is what has worked for me, I don't know if there is a better way that avoids this.

**Comment by naNEQ** [comment]
I must say that I have now decided to switch back to the upstream driver. I think it more or less works fine with ADS1015, one just needs to know that the driver assumes there is only one channel. This is in essence correct, as the other channels are "artificially" created with the MUX, but the ADC core has one channel. Thus, the user only needs to define the device in the devicetree and then before each conversion, the ADC needs to setup the correct channel through the MUX (the driver is up for the job, the user needs to use `adc_channel_setup`.

```
ads1015_adc: ti_ads1015@48{
        compatible = "ti,ads1015";
        status = "okay";
        reg = <0x48>;
        #io-channel-cells = <1>;
    };
```

**Reply by GastonMelo** [reply]
Hi again, I was able to get the driver working and is detected by zephyr. Now I have another question....(yeahh again!!).
How to use it in the main.c? I read the ADC API and there is 3 ways to reference the devicetree in the main (using channel_cfg and dt_spec) and i see no example adc in the zephyr folder. If I use the dt_spec i need to create a node in the device tree with user,zephyr referencing to adc and channel, but if i use the channel_conf api? Is it me or the zephyr documentation lacks of information?
Gastón

**Reply by GastonMelo** [reply]
Im struggling with the setup!! I got the driver enabled but I tried to read a sample for the ads1100, in the main I got this:

`
static const struct adc_dt_spec ads1100 =
    ADC_DT_SPEC_GET_BY_IDX(DT_PATH(zephyr_user), 0);

    if(!adc_is_ready_dt(&ads1100))
    {
        LOG_ERR("device not ready");
        return;
    }    
    ret=adc_channel_setup_dt(&ads1100);
    {
        LOG_ERR("could not setup device",0,ret);
        return;
    }
`

and I only got the error mesagge about "could not setup device". But why?

In the overlay i got this:

`
/
{
    zephyr,user {
        io-channels = <&ads1100 0>;
    };
};

&i2c0 {   
	ads1100: ads1100@48 {
		status = "okay";
		compatible = "ti,ads1100";
		reg = <0x48>;
		#io-channel-cells = < 1 >;
		#address-cells = < 1 >;
		#size-cells = < 0 >;
		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_VDD_1";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;            
		};
    };
`
Gastón

**Reply by naNEQ** [reply]
> Hi again, I was able to get the driver working and is detected by zephyr. Now I have another question....(yeahh again!!). How to use it in the main.c? I read the ADC API and there is 3 ways to reference the devicetree in the main (using channel_cfg and dt_spec) and i see no example adc in the zephyr folder. If I use the dt_spec i need to create a node in the device tree with user,zephyr referencing to adc and channel, but if i use the channel_conf api? Is it me or the zephyr documentation lacks of information? Gastón

@GastonMelo in the latest release 3.7 there was actually released a new sample which uses `adc_channel_setup`. This function doesn't rely on information from the device tree. Have a look there. I personally think this way it is easier to apply the desired configuration. Here is the [link](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/adc/adc_sequence/src/main.c) to the sample.

Also, another tip from my side which has saved me many times. If a sample (example) is not readily available for what you are looking for, do a general search on all subfolders of the Zephyr repo, e.g., in your case search for the function `adc_channel_setup`. Chances are high that what you're looking for is used somewhere else.

**Comment by Chengyen851010** [comment]
Hi everyone, I’m currently working on reading data from a pressure sensor using Zephyr, an ADS1115 ADC, and a Raspberry Pi 4. I’ve gone through all the responses in this thread and set up my environment accordingly, but unfortunately, I still can’t get the device to be detected. Would anyone be able to help me troubleshoot or point out what I might be missing? I’d really appreciate any guidance. Thanks!

## overlay
I add an overlay file here `/zephyr/boards/raspberrypi/rpi_4b/rpi_4b.overlay`
```
/ {
	zephyr,user {
		io-channels = <&ads1115 0>;
	};
};

&i2c2 {
	status = "okay";
	ads1115: ads1115@48 {
		status = "okay";
		compatible = "ti,ads1115";
		reg = <0x48>;
		#io-channel-cells = < 1 >;
		#address-cells = < 1 >;
		#size-cells = < 0 >;

		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
		};
	};
};
```

## prj.conf
```
CONFIG_ADC=y
CONFIG_LOG=y
CONFIG_GPIO=y
CONFIG_I2C=y

CONFIG_SERIAL=y
CONFIG_CONSOLE=y
CONFIG_PRINTK=y
CONFIG_UART_CONSOLE=y
CONFIG_UART_LINE_CTRL=y
CONFIG_UART_ASYNC_API=y
CONFIG_ADC_ADS1X1X=y


CONFIG_PICOLIBC_USE_MODULE=y
CONFIG_POSIX_API=y

```

## C 
```c
  #include <zephyr/kernel.h>
  #include <zephyr/sys/printk.h>
  #include <zephyr/drivers/i2c.h>
  #include <zephyr/drivers/adc.h>
  const struct device *adc_dev = DEVICE_DT_GET_ANY(ads1115);
  int main(void) {
    if (adc_dev == NULL) {
      printk("Failed to get ADS1115 ADC device binding\n");
      return 0;
    }
    return 0;
  };

```

## Result
```
Failed to get ADS1115 ADC device binding
```

**Reply by naNEQ** [reply]
Can you elaborate a bit more what it means that you can't get the device to be detected? What exactly is the error message? Is it during building or during runtime that you see the error?

**Reply by boaks** [reply]
For a

```
compatible = "ti,ads1115";
```

in the device tree you must use

```
const struct device *adc_dev = DEVICE_DT_GET_ANY(ti_ads1115);
```

to access it. "ti,ads1115" is converted into "ti_ads1115"

**Comment by boaks** [comment]
There is an other issue, #78973, which explains the background.

As midterm work-around, I use an approach with [multiple standard channel definitions](https://github.com/boaks/zephyr-coaps-client/blob/ads1115/ads1115.overlay#L44-L60).

```
ads1115: ads1115@48 {
		status = "okay";
		compatible = "ti,ads1115";
		reg = <0x48>;
		#io-channel-cells = <1>;
		#address-cells = < 1 >;
		#size-cells = < 0 >;

		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
			zephyr,input-positive = <0>;
		};

		channel@1 {
			reg = <1>;
			zephyr,gain = "ADC_GAIN_1_2";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = < 15 >;
			zephyr,input-positive = <1>;
		};

	};
```

As already written above, the driver will reject all channels with a different id as 0.
Therefore you need to [tweak](https://github.com/boaks/zephyr-coaps-client/blob/ads1115/src/appl_adc.c#L60-L70):

```
   struct adc_dt_spec adc_channel;

   adc_channel = adc_channels[channel];
   // overwrite logic channel id by 0
   adc_channel.channel_id = 0;
   adc_channel.channel_cfg.channel_id = 0;

   if (!adc_is_ready_dt(&adc_channel)) {
      LOG_INF("ADC controller device %s not ready\n", adc_channel.dev->name);
      return -EINVAL;
   }

   err = adc_channel_setup_dt(&adc_channel);

```

Not really nice and requires some extra Macros to be portable for other ADCs. But it allows to define the channels in the device tree.
